rpc: backpressured RPC server (bump jsonrpsee 0.20)

niklasad1 opened the pull request:
This is rather big change but most stuff are "chore", the major things in this bump are:

- Server backpressure (the subscription impls are modified to deal with that)
- Allow custom error types / return types (remove jsonrpsee::core::Error and jsonrpee::core::CallError)
- Easier subscription API (spawned internally by jsonrpsee doesn't work well with rpc subscription prometheus, so these are spawned with oneshot to keep it alive)
- Bug fixes (graceful shutdown in particular not used by substrate anyway)
- Less dependencies for the clients in particular
- Return type requires `Clone` in method call responses
- tokio channels are now used and tests must run on the tokio runtime to not panic

Hopefully the last release prior to 1.0, sorry in advance for a big PR

polkadot companion: https://github.com/paritytech/polkadot/pull/7211

cumulus companion: https://github.com/paritytech/cumulus/pull/2560

niklasad1 committed:
companion for jsonrpsee v0.17

niklasad1 committed:
dummy debug

niklasad1 committed:
rewrite me

niklasad1 committed:
adjust to custom subscription close type

niklasad1 committed:
Merge remote-tracking branch 'origin/na-jsonrpsee-v0.17-dev' into na-jsonrpsee-v0.17-dev

niklasad1 committed:
hack hack

niklasad1 committed:
update jsonrpsee

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.17-dev

niklasad1 committed:
jsonrpsee v0.18

niklasad1 commented on client/rpc-servers/src/lib.rs:
```diff
@@ -92,6 +92,8 @@ pub async fn start_http<M: Send + Sync + 'static>(
 ) -> Result<ServerHandle, Box<dyn StdError + Send + Sync>> {
 	let max_payload_in = payload_size_or_default(max_payload_in_mb) as u32;
 	let max_payload_out = payload_size_or_default(max_payload_out_mb) as u32;
+	log::info!("max_payload_out: {}, max_payload_in: {}", max_payload_out, max_payload_in);
```
```suggestion
```

niklasad1 committed:
Update client/rpc-servers/src/lib.rs

niklasad1 committed:
Update client/rpc-servers/src/lib.rs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2722608

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2722609

niklasad1 committed:
cleanup

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726288

niklasad1 committed:
Update client/rpc/src/chain/chain_full.rs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726436

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726429

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726431

niklasad1 committed:
fix rpcv2 tests

niklasad1 committed:
Merge remote-tracking branch 'origin/na-jsonrpsee-v0.18' into na-jsonrpsee-v0.18

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2731615

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2731614

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2731613

niklasad1 committed:
fix compile warns

niklasad1 committed:
update jsonrpsee branch

niklasad1 committed:
update jsonrpsee v0.18.1

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2744882

niklasad1 committed:
remove ugly todo's

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2745056

niklasad1 committed:
use subscription task executor

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2748829

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

niklasad1 committed:
update remote externalities

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2749398

niklasad1 committed:
fix nits

niklasad1 commented on client/rpc/src/lib.rs:
```diff
@@ -43,3 +43,121 @@ pub mod testing;
+		T: Serialize + Send + 'static,
+		R: Serialize + Send + 'static,
+	{
+		let Ok(sink )= pending.accept().await else {
```
```suggestion
		let Ok(sink) = pending.accept().await else {
```

niklasad1 committed:
Update client/rpc/src/lib.rs

niklasad1 commented on client/rpc/src/lib.rs:
```diff
@@ -43,3 +43,121 @@ pub mod testing;
+							None => break SubscriptionResponse::Closed,
+						};
+
+						match sink.send_timeout(msg, std::time::Duration::from_secs(60)).await {
```
The major change is here: if the subscription message could not be sent in 60 seconds => drop the subscription

This is really naive and the new spec has specified behavior when to drop etc.. but keep somewhat "backward compatible API".

We could also send a special message when this occurs but the client is most likely disconnected anyway..

niklasad1 commented on client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:
```diff
@@ -528,62 +530,47 @@ where
-					return
-				}
+				match sink
+					.send_timeout(sc_rpc::utils::to_sub_message(&event), Duration::from_secs(60))
```
//cc @lexnv what to reckon to do here?

lexnv replied:
This should be fine for chainHead use cases.
In the end, we want to backpressure connections that are not responsive to avoid wasting resources, at the same time 60 seconds should be plenty of time.
Further, the spec states that the node can submit the `Stop` event when it is overloaded, I interpret that to be the case here :D

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2750070

niklasad1 committed:
fix tests

niklasad1 committed:
Merge remote-tracking branch 'origin/na-jsonrpsee-v0.18' into na-jsonrpsee-v0.18

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

niklasad1 committed:
fix tests

niklasad1 committed:
make it possible to configure buf capacity

liamaharon commented on client/cli/src/config.rs:
```diff
@@ -53,8 +53,11 @@ pub const RPC_DEFAULT_MAX_SUBS_PER_CONN: u32 = 1024;
 /// The default max response size in MB.
 pub const RPC_DEFAULT_MAX_RESPONSE_SIZE_MB: u32 = 15;
-/// The default number of connection..
+/// The default number of connections.
```
```suggestion
/// The default concurrent connection limit.
```

liamaharon commented on client/cli/src/config.rs:
```diff
@@ -346,6 +349,11 @@ pub trait CliConfiguration<DCV: DefaultConfigurationValues = ()>: Sized {
 		Ok(RPC_DEFAULT_MAX_SUBS_PER_CONN)
 	}
 
+	/// Get maximum number of subscriptions per connection.
```
Is this only subscriptions? or messages generally?

maybe:

```suggestion
	/// The number of messages the RPC server is allowed to keep in memory per connection.
```

niklasad1 replied:
messages, thanks

bad copy-paste nit

liamaharon commented on client/cli/src/config.rs:
```diff
@@ -53,8 +53,11 @@ pub const RPC_DEFAULT_MAX_SUBS_PER_CONN: u32 = 1024;
-/// The default number of connection..
+/// The default number of connections.
 pub const RPC_DEFAULT_MAX_CONNECTIONS: u32 = 100;
+/// The default number messages the RPC server
+/// are allowed to keep in memory until.
```
```suggestion
/// The default number of messages the RPC server
/// is allowed to keep in memory per connection.
```

liamaharon commented on client/rpc-api/src/system/error.rs:
```diff
@@ -45,17 +51,16 @@ const NOT_HEALTHY_ERROR: i32 = BASE_ERROR + 1;
-				None::<()>,
-			)),
+				ErrorObject::owned(NOT_HEALTHY_ERROR, e.to_string(), Some(h)),
+			Error::MalformattedPeerArg(e) =>
+				ErrorObject::owned(MALFORMATTED_PEER_ARG_ERROR + 2, e, None::<()>),
```
`MALFORMATTED_PEER_ARG_ERROR` is already `BASE_ERROR + 2;` on ln 52.

Just checking it is intentional to bump it another `2` here.

niklasad1 replied:
good catch thanks,

liamaharon commented on client/rpc-spec-v2/src/chain_head/tests.rs:
```diff
@@ -885,8 +913,8 @@ async fn follow_with_unpin() {
 	let block3 = client.new_block(Default::default()).unwrap().build().unwrap().block;
 	client.import(BlockOrigin::Own, block3.clone()).await.unwrap();
 
-	assert_matches!(get_next_event::<FollowEvent<String>>(&mut sub).await, FollowEvent::Stop);
-	assert!(sub.next::<FollowEvent<String>>().await.is_none());
+	//assert_matches!(get_next_event::<FollowEvent<String>>(&mut sub).await, FollowEvent::Stop);
+	//assert!(sub.next::<FollowEvent<String>>().await.is_none());
```
intentionally commented?

liamaharon commented on client/rpc/src/lib.rs:
```diff
@@ -44,3 +44,124 @@ pub mod testing;
+		pipe_from_stream(sink, stream, executor).await
+	}
+
+	/// Feed items to the subscription from the underlying stream
+	/// if the subscription can't keep up with the underlying stream
+	/// it's dropped
```
```suggestion
	/// Feed items to the subscription from the underlying stream.
	/// If the subscription can't keep up with the underlying stream
	/// then it's dropped.
```

niklasad1 committed:
Update client/cli/src/config.rs

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>

niklasad1 committed:
Update client/rpc/src/lib.rs

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>

niklasad1 committed:
Update client/cli/src/config.rs

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>

niklasad1 committed:
Update client/cli/src/config.rs

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2810824

niklasad1 committed:
address grumbles

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2811075

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2811074

niklasad1 commented on client/rpc/src/testing.rs:
```diff
@@ -20,11 +20,50 @@
+///
+/// Warning: the tokio runtime must be initialized before calling this.
+#[derive(Clone)]
+pub struct TokioTestExecutor(tokio::runtime::Handle);
```
jsonrpsee is now using tokio channels so we must run the tests on the tokio runtime to work..

niklasad1 commented on client/service/src/lib.rs:
```diff
@@ -111,9 +111,9 @@ impl RpcHandlers {
+	) -> Result<(String, tokio::sync::mpsc::Receiver<String>), JsonRpseeError> {
 		self.0
-			.raw_json_request(json_query)
+			.raw_json_request(json_query, usize::MAX)
```
this was unbounded before so let's keep it like that.

koute committed:
Remove the `Copy` bound on `CollectionId` in the uniques pallet (#14111)

* Remove the `Copy` bound on `CollectionId` in the uniques pallet

* Also add `clone`s in benchmarks

gavofyork committed:
Timeout only if the referendum is not queued (#14106)

* Timeout only if the referendum is not queued

* ".git/.scripts/commands/fmt/fmt.sh"

---------

Co-authored-by: command-bot <>

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

ggwpez committed:
Include `node-template-release` in workspace (#14103)

* Add node-template-release to workspace

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Remove empty workspace

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Manually update deps

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update Cargo.lock

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update .gitignore

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update license

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

tdimitrov committed:
Bump `kvdb-rocksdb` to 0.19.0 (#14113)

ggwpez committed:
Remove `#[pallet::generate_storage_info]` from docs (#14116)

* Fix docs

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Add UI test

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

rcny committed:
Bump `Cargo.lock` (#14121)

* Bump `Cargo.lock`

* Use master Cargo.lock

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update Cargo.lock

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

skunert reviewed (approved):
Looks good!

altonen committed:
Prepare `sc-network` for `ProtocolController`/`NotificationService` (#14080)

* Prepare `sc-network` for `ProtocolController`/`NotificationService`

The upcoming notification protocol refactoring requires that protocols
are able to communicate with `sc-network` over unique and direct links.
This means that `sc-network` side of the link has to be created before
`sc-network` is initialized and that it is allowed to consume the object
as the receiver half of the link may not implement `Clone`.

Remove request-response and notification protocols from `NetworkConfiguration`
and create a new object that contains the configurations of these protocols
and which is consumable by `sc-network`. This is needed needed because, e.g.,
the receiver half of `NotificationService` is not clonable so `sc-network`
must consume it when it's initializing the protocols in `Notifications`.

Similar principe applies to `PeerStore`/`ProtocolController`: as per current
design, protocols are created before the network so `Protocol` cannot be
the one creating the `PeerStore` object. `FullNetworkConfiguration` will be
used to store the objects that `sc-network` will use to communicate with
protocols and it will also allow protocols to allocate handles so they
can directly communicate with `sc-network`.

* Fixes

* Update client/service/src/builder.rs

Co-authored-by: Dmitry Markin <dmitry@markin.tech>

* Updates

* Doc updates + cargo-fmt

---------

Co-authored-by: Dmitry Markin <dmitry@markin.tech>

niklasad1 committed:
jsonrpsee v0.18.2

niklasad1 commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

niklasad1 commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

niklasad1 commented:
this is blocked on a pending burn-in

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

niklasad1 committed:
pin jsonrpsee to commit

niklasad1 committed:
decrease message buffer to 16

niklasad1 committed:
Merge remote-tracking branch 'origin/na-jsonrpsee-v0.18' into na-jsonrpsee-v0.18

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2852897

niklasad1 committed:
rpc helper: unwrap -> unwrap_or

niklasad1 committed:
update jsonrpsee

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2917450

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2917451

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

niklasad1 committed:
rpc: subscribe_storage use spawn_blocking

niklasad1 committed:
change default rpc message buffer

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3015154

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3015297

niklasad1 committed:
fix nit

niklasad1 committed:
add some debug logs

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3064656

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3064655

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-jsonrpsee-v0.18

niklasad1 committed:
update jsonrpsee

niklasad1 committed:
revert me

niklasad1 committed:
update jsonrpsee

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3099078

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3099079

niklasad1 committed:
jsonrpsee v0.19

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3295104

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3295105

kayabaNerve commented:
Sorry for walking in without context, yet what were the regressions identified preventing this from being merged?

niklasad1 commented:
> Sorry for walking in without context, yet what were the regressions identified preventing this from being merged?

A leak and that the new async subscription API allocates more memory.
